staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go

```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+)
+
+// WithRetryAfter rejects any incoming new request(s) with a 429
+// if the specified shutdownDelayDurationElapsedCh channel is closed
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after shutdownDelayDurationElapsedCh is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
```

Comment:
This filter changes the status codes previously returned on paths like `livez` and `readyz` previously `500` now it will be `429`. Again this might be not a backward compatible change/behavriour.

Reply:
We might actually allow some actors to get through like `the loopback client`. Otherwise, the server might not shut down gracefully.

Reply:
readyz/livez/healthyz/metrics definitely should not change

Reply:
+1 for loopback client

Reply:
I updated the filter to check the path and the user, PTAL https://github.com/kubernetes/kubernetes/pull/103357

Reply:
the `shutdownDelayDurationElapsedCh` is a detail of the caller, not the func. Would call the func `WithRetryAfterDuringShutdown(handler, shutdownCh)`. Then it is understandable without context.